### ISSUE TITLE
Integration with localised data (using react-i18next & i18next)

### ISSUE THREAD

**Issue by leimonio** [nozzle/react-static#475]
I'm migrating an existing non-react application which loads the locales from local files. What I'd like to do is to load this localized data from the files using `react-i18next` and `i18next`. There is not any related documentation on your website, and what I'd like to know is which is the best and proposed way to integrate localized data with the building of the static pages with `react-static`. The documentation mentions that all the needed data must be loaded in `static.config.js`, but `react-i18next` uses a different configuration using the context to provide with the localized data in the components.

So far I have updated my `App.js` as:
```
<I18nextProvider i18n={i18n}>
  <Router>
    <Routes />
  </Router>
</I18nextProvider>
```

**Comment by JustFly1984** [comment]
I have my React app internationalization made with `react-intl`, and it works like a charm. If you need help, ping me in skype: justfly1984

**Comment by tannerlinsley** [comment]
Yep, this is perfectly valid way of doing internationalization. There are some extra things to consider like a default language for static exports and also exporting each language. This has been done by a few people, so I would suggest contacting @JustFly1984 or even jumping on the slack channel for more help on implementation :)

**Comment by rohmanhm** [comment]
Hi @leimonio, I have an app that uses `react-static` and `react-i18next` and it works really well.

To integrate, you just need to edit i18n instance.
Just look the code below.
Hope it helps

```ts
import i18next from 'i18next'
import Backend from 'i18next-xhr-backend'
import LanguageDetector from 'i18next-browser-languagedetector'
import { reactI18nextModule } from 'react-i18next'

const options = {
  fallbackLng: 'en',

  // have a common namespace used around the full app
  ns: ['home'],
  defaultNS: 'home',

  debug: process.env.REACT_STATIC_ENV !== 'production',

  interpolation: {
    escapeValue: false, // not needed for react!!
  },

  react: {
    wait: true
  }
}


const i18n = i18next

// for browser use xhr backend to load translations and browser lng detector
if (typeof document !== 'undefined') {
  i18n
    .use(Backend)
    .use(LanguageDetector)
}

// initialize if not already initialized
if (!i18n.isInitialized) {
  i18n
    .use(reactI18nextModule)
    .init(options)
}

export default i18n
```

**Comment by AlexFBP** [comment]
Is there a way to trully generate react-static content? I mean, rather than lying on browser detection and XHR locale loading; using path based translations, overall if the project uses `@reach/router`
as example: that `example.com` loads the fallback language and `example.com/es` loads locale for es

**Comment by klis87** [comment]
+1 @AlexFBP , this is especially important for SEO, different languages should be available on separate paths, also, coming to `example.com/es` should get index.html with already translated words to Spanish.

**Comment by tannerlinsley** [comment]
You can create as many route variations you want using `config.getRoutes()`:

```javascript
const languages = ['en', 'es']

export default {
  getRoutes: () => {
  	const routes = []
    languages.forEach(language => {
    	routes = [
        	...routes,
            ...[{
        		path: `${language}/about`,
            	component: 'path-to-component.js'
        	}, {
        		path: `${language}/blog`,
            	component: 'path-to-component.js'
        	}]
          ]
    })
    return routes
  }
}
```

**Comment by klis87** [comment]
@tannerlinsley thx, I have still one puzzle missing though, in `App.js` I wanted to get current language, but it show loading spinner all the time. I tried this:
```js
import React from 'react'
import { Root, Routes, RouteData } from 'react-static'
import { Link } from '@reach/router'

import './app.css'

function App() {
  return (
    <Root>
      <nav>
        <Link to="/">Home</Link>
        <Link to="/about">About</Link>
        <Link to="/blog">Blog</Link>
      </nav>
      <div className="content">
        <RouteData>
          {({ lang }) => (
            <div>
              {lang}
            </div>
          )}
        </RouteData>
        <Routes />
      </div>
    </Root>
  )
}

export default App
```
but `RouteData` shows only spinner all the time. I also tried to use render prop with getComponentForPath in Routes, but `lang` prop is not available there as well. I need this to know which language I should use to inject i18next provider with proper language. Also, really this provider should wrap header too, because menu items need to be translated as well. Is it possible somehow?

**Comment by tannerlinsley** [comment]
Assuming you pass the `language` down to every route via `getData`, you can use it like this:

```js
const languages = ['en', 'es']

export default {
  getRoutes: () => {
  	let routes = []
    languages.forEach(language => {
    	routes = [
        ...routes,
        ...[{
          path: `${language}/about`,
          component: 'path-to-component.js'
          getData: () => ({ language })
        }, {
          path: `${language}/blog`,
          component: 'path-to-component.js'
          getData: () => ({ language })
        }]
      ]
    })
    return routes
  }
}

// in your application...

<RouteData>
  {({ language }) => (
    <div>The language is: {language}</div>
  )}
</RouteData>
```

**Comment by klis87** [comment]
@tannerlinsley hmm, In my example I tried to do it this way but for some reason I didnt work, I dont know the reason but what matters is that it works now, thanks!.

If anyone has issue with it too, note that u have to nest RouteData inside Routes, here is the snippet:
```js
function App() {
  return (
    <Root>
      <Routes>
        {({ getComponentForPath }) => {
          const Comp = getComponentForPath(window.location.href);

          return (
            <>
              <RouteData>
                {({ lang }) => <div>language: {lang}}</div>}
              </RouteData>
              <nav>
                <Link to="/">Home</Link>
                <Link to="/about">About</Link>
                <Link to="/blog">Blog</Link>
              </nav>
              <div className="content">
                <Comp />
              </div>
            </>
          );
        }}
      </Routes>
    </Root>
  );
}
```

**Comment by klis87** [comment]
I spent quite amount of time to connect all the pieces together. It works but there are some things could be better. I will mention them, if you want, I could raise dedicated issues too later, but see my implementation first:
```js
// static.config.js
const languages = ['pl', 'en'];

export default {
  plugins: ['react-static-plugin-react-router'],
  getSiteData: () => ({
    title: 'React Static',
  }),
  minLoadTime: 1,
  getRoutes: () => {
    const routes = [
      {
        path: '/',
        redirect: '/pl',
      },
    ];
    languages.forEach(lang => {
      routes.push(
        {
          path: `/${lang}/`,
          component: 'src/containers/index',
          getData: () => ({
            lang,
          }),
        },
        {
          path: `/${lang}/about/`,
          component: 'src/containers/about',
          getData: () => ({
            lang,
          }),
        },
      );
    });

    return routes;
  },
  maxThreads: 1, // not to create multiple i18n copies
};
```

```js
// node.api.js
import React from 'react';
import i18next from 'i18next';
import backend from 'i18next-sync-fs-backend';

i18next.use(backend).init({
  preload: ['pl', 'en'],
  fallbackLng: false,
  debug: false,
  load: 'languageOnly',
  ns: ['common'],
  defaultNS: 'common',
  initImmediate: false,
  backend: { loadPath: 'public/locales/{{lng}}/{{ns}}.json' },
  react: {
    wait: false,
    // useSuspense: false,
    bindStore: false,
    bindI18n: 'languageChanged',
  },
});

const getClonedInstance = () => {
  const instance = i18next.cloneInstance();
  return instance;
};

export default () => ({
  beforeRenderToElement: (App, { meta }) => {
    meta.i18next = getClonedInstance();
    return props => <App {...props} serverI18next={meta.i18next} />;
  },
  Head: ({ meta }) => {
    const language = meta.i18next.language || 'pl';
    const store = {
      [language]: meta.i18next.services.resourceStore.data[language],
    };
    
   // for some reason key is needed to avoid react warning during build
    return (
      <script
        key="initialI18nStore"
        dangerouslySetInnerHTML={{
          __html: `window.initialI18nStore =  '${JSON.stringify(store)}';`,
        }}
      />
    );
  },
});
```

```js
// i18n provider wrapper used in App.js
import clientI18next from 'i18next';
import { I18nextProvider } from 'react-i18next';
import React from 'react';

if (typeof window !== 'undefined') {
  clientI18next.use(require('i18next-xhr-backend')).init({
    fallbackLng: false,
    debug: process.env.REACT_STATIC_ENV === 'development',
    load: 'languageOnly',
    ns: ['common'],
    defaultNS: 'common',
    backend: { loadPath: '/locales/{{lng}}/{{ns}}.json' },
    react: {
      wait: false,
      // useSuspense: false, // for newest react-i18next
      bindStore: false,
      bindI18n: 'languageChanged',
    },
  });
}

const I18nProvider = ({ children, lang, serverI18next }) => {
  if (typeof window === 'undefined') {
    serverI18next.changeLanguage(lang || 'pl'); // || pl needed for 404
  }

  return (
    <I18nextProvider
      i18n={serverI18next || clientI18next}
      initialLanguage={typeof window !== 'undefined' ? lang : undefined}
      initialI18nStore={
        typeof window !== 'undefined' && window.initialI18nStore
          ? JSON.parse(window.initialI18nStore)
          : undefined
      }
    >
      {children}
    </I18nextProvider>
  );
};
```

Problems?
1) node.api.js isn't executed in development, so window.initialI18nStore is not defined there, so during development we have flickering
2) There is no way to wait until all languages are preloaded, that's why I used sync fs backend on node, not big deal but this is a thing to think about whether some async await hook to prepare a plugin could be added to this library
3) weird `key` issue in `Head`, is it intended?
4) pity that `beforeRenderToElement` is not executed for browser, hence I needed to init i18n for client on the app level. Maybe `browser.api.js` could allow to wrap App in similar way to add extra props too like `beforeRenderToElement`?
5) there is really no way to know in `beforeRenderToElement` which language should be picked for a given route, as I don't have access there to data and path. Hence I need to pass i18n as prop and call `serverI18next.changeLanguage` in render which is really ugly. If I had access to language in `beforeRenderToElement`, I could even put i18n provider in plugins, so app wouldn't need to do it anymore
6) handling of multiple namespaces is yet to be added, but this should be easy even with current library version, to put only relevant namespaces to `window.initialI18nStore`, it is not react-static issue but I am writing it for completion

To sum up, if above points were fixed, it would be possible to create really nice i18next plugin for react static, as well as many more integrations we didn't think of. I checked some plugins in this library like jss, emotions, and current API was enough for those. Unfortunately more complex cases like i18next need a little more flexibility

**Comment by Geoff-Ford** [comment]
Hi @klis87 Did you get any further on with this? Trying to implement the `react-i18next` and also use a url param for lang.

**Comment by klis87** [comment]
@Geoff-Ford yes, but sadly not as integration encapsulated in the plugin, and I did it with older i18next (without React hooks). Some stuff is integrated outside, like I described above. If u have any issues I will give you some snippets later.

**Comment by chiqui3d** [comment]
I've make it work in `dev `mode, but when I do the `build`, it throws an error. An example of how to apply `i18next `it would be very helpful.
